### PR TITLE
Fixed point arithmetic

### PR DESCRIPTION
**Pull request by jonas-lj**
Added functionality for fixed point operations and linear algebra.

### CODE REVIEW

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  public DRes<SReal> fromSInt(DRes<SInt> value) {
+    return builder.seq(seq -> {
+      return new SFixed(value.out(), 0);
+    });
+  }
+
+  @Override
+  public DRes<SReal> input(BigDecimal value, int inputParty) {
+    return builder.seq(seq -> {
+      int precision = defaultPrecision;
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java]: This variable does not make sense.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+      DRes<SInt> scaledResult = seq.advancedNumeric().div(aInt, bUnscaled);
+      return new SFixed(scaledResult, defaultPrecision);
+    });
+  }
+
+  @Override
+  public DRes<SReal> known(BigDecimal value) {
+    return builder.seq(seq -> {
+      int precision = defaultPrecision;
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java]: This variable does not make sense.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/FixedLinearAlgebra.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package dk.alexandra.fresco.decimal.fixed;
+
+import dk.alexandra.fresco.decimal.DefaultLinearAlgebra;
+import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
+
+public class FixedLinearAlgebra extends DefaultLinearAlgebra {
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/FixedLinearAlgebra.java]: This class is unneeded. Remove it.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/SFixed.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import dk.alexandra.fresco.decimal.SReal;
+import dk.alexandra.fresco.framework.DRes;
+import dk.alexandra.fresco.framework.value.SInt;
+
+/**
+ * Closed datatype for representing binary fixed point numbers, e.g. represent a fraction <i>x</i>
+ * as <i>n 2<sup>e</sup></i> where <i>n</i> is an {@link SInt}, <i>e &ge; 0</i> is a precision
+ * (avaialble via {@link #getScale()}) that may vary from value to value.
+ * 
+ * @author Jonas Lindstrøm (jonas.lindstrom@alexandra.dk)
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/SFixed.java]: author..

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/SFixed.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ */
+public class SFixed implements SReal, DRes<SReal> {
+  private final DRes<SInt> value;
+  private int scale;
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/SFixed.java]: final

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/utils/Truncate.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import dk.alexandra.fresco.framework.builder.numeric.AdvancedNumeric.RandomAdditiveMask;
+import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
+import dk.alexandra.fresco.framework.util.Pair;
+import dk.alexandra.fresco.framework.value.SInt;
+import dk.alexandra.fresco.lib.math.integer.binary.RightShift;
+import java.math.BigInteger;
+
+/**
+ * Returns a number which is approximately the input shifted a number of positions to the right. The
+ * result will be one larger than the exact result with some non negligeble propability. If you need
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/utils/Truncate.java]: non negligeble
is in need of an a and a dash (-)

File: tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/BasicFixedPointTests.java [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+import org.junit.Assert;
+
+public class BasicFixedPointTests {
+
+  public static class TestInput<ResourcePoolT extends ResourcePool>
+      extends TestThreadFactory<ResourcePoolT, ProtocolBuilderNumeric> {
+
+    @Override
+    public TestThread<ResourcePoolT, ProtocolBuilderNumeric> next() {
+      BigDecimal value = BigDecimal.valueOf(10.00100);
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/BasicFixedPointTests.java]: Please use _ in long numbers
https://docs.oracle.com/javase/8/docs/technotes/guides/language/underscores-literals.html

File: tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/BasicFixedPointTests.java [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.Assert;
+
+public class BasicFixedPointTests {
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/BasicFixedPointTests.java]: In general a lot of these test only run on a single number. It would be very reasonable to include a set of dimensions (number of decimals, edge cases, moduli size, ...) and test each combination

File: tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/LinearAlgebraTests.java [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+  public static class TestMatrixAddition<ResourcePoolT extends ResourcePool>
+      extends TestThreadFactory<ResourcePoolT, ProtocolBuilderNumeric> {
+
+    @Override
+    public TestThread<ResourcePoolT, ProtocolBuilderNumeric> next() {
+      return new TestThread<ResourcePoolT, ProtocolBuilderNumeric>() {
+
+        @Override
+        public void test() throws Exception {
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/test/java/dk/alexandra/fresco/fixedpoint/basic/LinearAlgebraTests.java]: Similarly these test could benefit from being produced randomly, with a focus on edge cases

**Comment by pffrandsen** [comment]
One comment - why not include it in core?
In the current construct we now have to include both spdz and real numbers

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/AdvancedRealNumeric.java [hunk 10]
```diff
@@ -10,34 +10,34 @@
...
   DRes<SReal> innerProduct(List<DRes<SReal>> a, List<DRes<SReal>> b);
 
   /**
    * Calculate the inner product of a public and a secret vector.
    * 
-   * @param a Public value
-   * @param b Secret value
-   * @return
+   * @param a List of public values
+   * @param b List of secret values
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/AdvancedRealNumeric.java]: how about changing the parameters to say, publicValues and secretValues?

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/LinearAlgebra.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  DRes<Matrix<DRes<SReal>>> mult(DRes<Matrix<DRes<SReal>>> a, DRes<Matrix<DRes<SReal>>> b);
+
+  /**
+   * Multiply a matrix to a vector.
+   * 
+   * @param a Secret value 1
+   * @param v Secret value 2
+   * @return A deferred result computing a*v
+   */
+  DRes<Vector<DRes<SReal>>> operate(DRes<Matrix<DRes<SReal>>> a, DRes<Vector<DRes<SReal>>> v);
```
Comment (pffrandsen) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/LinearAlgebra.java]: Let us call this method vectorMult? operate is a very bad name

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/DefaultAdvancedRealNumeric.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+      return () -> new Pair<>(terms, divisor);
+    }).seq((seq, termsAndDivisor) -> {
+      RealNumeric fixed = provider.apply(seq);
+      return fixed.numeric().div(fixed.advanced().sum(termsAndDivisor.getFirst()),
+          termsAndDivisor.getSecond());
+    });
+  }
+
+  @Override
+  public DRes<SReal> random() {
```
Comment (GuutBoy) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/DefaultAdvancedRealNumeric.java]: This looks like a very dangerous way to implement the random method: 
1. The `DEFAULT_RANDOM_BITS` of 32 bits is not very much in a cryptographic setting  
2. This is not documented in the Javadoc of the API, so there is not way for a user to know that this is the case. The javadoc just says a "random number between 0 and 1". This should be made much more precise so the user knows how much entropy to except.
3. Implementing it using a random additive mask is very costly as each bit requires at least a single triple in the preprocessing phase.

While 3. may be unavoidable, 1. and 2. really bad and could very well damage the security of computations build using this method.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/RealNumeric.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package dk.alexandra.fresco.decimal;
+
+public interface RealNumeric {
```
Comment (GuutBoy) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/RealNumeric.java]: I think this naming is a little confusing. I would expect RealNumeric to be a counterpart to Numeric, but that does not seem to be the case.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    this.maxPrecision = builder.getBasicNumericContext().getMaxBitLength() / 4;
+    if (maxPrecision < 2 * defaultPrecision) {
+      throw new IllegalArgumentException(
+          "The precision was chosen too large for a product of two numbers to be representable "
+              + "in this numeric context. You cannot choose a precision larger than "
+              + (maxPrecision / 2) + ".");
+    }
+  }
+
+  public BasicFixedNumeric(ProtocolBuilderNumeric builder) {
```
Comment (GuutBoy) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java]: We should remember to document these constructors, in this case particularly the default precision used need to be known by the user of this class.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+import dk.alexandra.fresco.decimal.SReal;
+import dk.alexandra.fresco.decimal.fixed.utils.Truncate;
+import dk.alexandra.fresco.framework.DRes;
+import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
+import dk.alexandra.fresco.framework.value.SInt;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.RoundingMode;
+
+public class BasicFixedNumeric implements BasicRealNumeric {
```
Comment (GuutBoy) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/BasicFixedNumeric.java]: Generally this is missing javadoc.

File: tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/FixedNumeric.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
+
+/**
+ * This implementation of {@link RealNumeric} allows numbers to be represented as binary fixed point
+ * numbers, eg. as <i>m 2<sup>-n</sup></i> for integers <i>m, n</i> with <i>0 &le; n &le; N</i> for
+ * some upper bound <i>N</i>. Inputs are converted to such a representation for a given default
+ * precision (see {@link #FixedNumeric(ProtocolBuilderNumeric, int)}), but intermediate results and
+ * outputs may be represented with higher precision.
+ * 
+ * @author Jonas Lindstrøm (jonas.lindstrom@alexandra.dk)
```
Comment (GuutBoy) [inline tools/fixedpointarithmetic/src/main/java/dk/alexandra/fresco/decimal/fixed/FixedNumeric.java]: Author tag.

**Comment by GuutBoy** [comment]
@jonas-lj I think one of the recent commits broke the code. It is no longer compiling.

**Comment by GuutBoy** [comment]
@jonas-lj Would you mind solving the merge conflics?

**Comment by pffrandsen** [comment]
@jonas-lj Fixed the merge, now we should at the coverage. There is no reason for not reaching 100 %

File: core/src/main/java/dk/alexandra/fresco/lib/real/AdvancedRealNumeric.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+   * inputs, so the input should be scaled if possible (note that <i>exp(n x) =
+   * exp(x)<sup>n</sup></i>).
+   * 
+   * @param x Secret value
+   * @return A deferred result computing computing e<sup>x</sup>
+   */
+  DRes<SReal> exp(DRes<SReal> x);
+
+  /**
+   * Create a random value between 0 and 1.
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/real/AdvancedRealNumeric.java]: This is still very problematic. We cannot sample such a number uniformly, so we need to document what can be expected of the sampling. Possibly, we should let the user indicate how many random bits should go into the sampling.

File: core/src/main/java/dk/alexandra/fresco/lib/real/fixed/AdvancedFixedNumeric.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+import dk.alexandra.fresco.framework.builder.numeric.AdvancedNumeric.RandomAdditiveMask;
+import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
+import dk.alexandra.fresco.framework.value.SInt;
+import dk.alexandra.fresco.lib.real.DefaultAdvancedRealNumeric;
+import dk.alexandra.fresco.lib.real.SReal;
+import java.math.BigDecimal;
+
+public class AdvancedFixedNumeric extends DefaultAdvancedRealNumeric {
+
+  private static final int DEFAULT_RANDOM_BITS = 64;
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/real/fixed/AdvancedFixedNumeric.java]: While this is in some sense better than before it does not solve the underlying problem. See comments on the `AdvancedRealNumeric` interface.

**Comment by GuutBoy** [comment]
About coverage you can see what is missing here: https://codecov.io/gh/aicis/fresco/pull/244/diff
It's pretty minor stuff in `FixedNumeric`, `DefaultAdvancedRealNumeric`, `DefaultLinearAlgebra` and `AdvancedFixedNumeric` that's left.
